internal/cumulative/cumulative.go

```diff
@@ -0,0 +1,138 @@
+package cumulative
+
+import (
+	"bytes"
+	"fmt"
+	pprofile "github.com/google/pprof/profile"
+	"github.com/pyroscope-io/client/upstream"
+	"time"
+)
+
+type ProfileMerger struct {
+	SampleTypes      []string
+	MergeRatios      []float64
+	SampleTypeConfig map[string]*upstream.SampleType
+
+	prev *pprofile.Profile
+}
+
+type Mergers struct {
+	Heap  *ProfileMerger
+	Block *ProfileMerger
+	Mutex *ProfileMerger
+}
+
+func NewMergers() *Mergers {
+	return &Mergers{
+		Block: &ProfileMerger{
+			SampleTypes: []string{"contentions", "delay"},
+			MergeRatios: []float64{-1, -1},
+			SampleTypeConfig: map[string]*upstream.SampleType{
+				"contentions": {
+					DisplayName: "block_count",
+					Units:       "lock_samples",
+				},
+				"delay": {
+					DisplayName: "block_duration",
+					Units:       "lock_nanoseconds",
+				},
+			},
+		},
+		Mutex: &ProfileMerger{
+			SampleTypes: []string{"contentions", "delay"},
+			MergeRatios: []float64{-1, -1},
+			SampleTypeConfig: map[string]*upstream.SampleType{
+				"contentions": {
+					DisplayName: "mutex_count",
+					Units:       "lock_samples",
+				},
+				"delay": {
+					DisplayName: "mutex_duration",
+					Units:       "lock_nanoseconds",
+				},
+			},
+		},
+		Heap: &ProfileMerger{
+			SampleTypes: []string{"alloc_objects", "alloc_space", "inuse_objects", "inuse_space"},
+			MergeRatios: []float64{-1, -1, 0, 0},
+			SampleTypeConfig: map[string]*upstream.SampleType{
+				"alloc_objects": {
+					Units: "objects",
+				},
+				"alloc_space": {
+					Units: "bytes",
+				},
+				"inuse_space": {
+					Units:       "bytes",
+					Aggregation: "average",
+				},
+				"inuse_objects": {
+					Units:       "objects",
+					Aggregation: "average",
+				},
+			},
+		},
+	}
+}
+
+// todo should we disableClientSideMerge if we fail to merge or keep trying?
+// todo should we filter by enabled ps.profileTypes to reduce profile size ? maybe add a separate option ?
+func (m *ProfileMerger) Merge(j *upstream.UploadJob) error {
```

Comment:
One potential issue with this code is that it is using the prev field of the ProfileMerger struct to store the previous profile. However, it is not clear from the code provided how this field is being initialized or whether it is being reset after each call to the Merge function. This could potentially cause issues if the prev field is not being properly initialized or reset, as it could lead to unexpected behavior when merging profiles.

Additionally, the code is using the defer keyword to print a message to the console after the Merge function has completed. This is indicated by the comment //todo remove before merge. It is generally a good idea to remove any debug code (such as this defer statement) before merging your code, as it can cause unexpected behavior and make it difficult to understand and debug your code.

Finally, the code is using the fmt.Printf function to print a message to the console. This function should not be used in production code, as it is intended for debugging purposes only. In production code, it is generally a better idea to use a logging library (such as the standard log package) to log messages. This allows you to easily control the log level and output location, which can be useful for debugging and monitoring your application.